gtsfm/utils/geometry_comparisons.py

```diff
@@ -30,138 +30,234 @@ def align_rotations(input_list: List[Rot3], ref_list: List[Rot3]) -> List[Rot3]:
+    input_pose_matrix = np.vstack([x.reshape(1, 3) for x in input_list])
+    ref_pose_matrix = np.vstack([x.reshape(1, 3) for x in ref_list])
+
+    # compute the mean pose
+    input_mean = np.mean(input_pose_matrix, axis=0)
+    ref_mean = np.mean(ref_pose_matrix, axis=0)
+
+    # subtract the means
+    input_pose_matrix = input_pose_matrix - input_mean.reshape(1, 3)
+    ref_pose_matrix = ref_pose_matrix - ref_mean.reshape(1, 3)
+
+    # compute the variances
+    input_variance = np.sum(np.square(input_pose_matrix), axis=None) / N
+    # ref_variance = np.sum(np.square(ref_pose_matrix), axis=None) / N
+
+    # compute the covariance matrix and perform SVD
+    cov_mat = np.dot(ref_pose_matrix.T, input_pose_matrix) / N
+    U, S, Vt = np.linalg.svd(cov_mat)
+    S = np.diag(S)
+
+    W = np.eye(3)
+    if np.linalg.det(U) * np.linalg.det(Vt) < 0:
+        W[2, 2] = -1
+
+    rot_matrix = U @ W @ Vt
+    scale = np.trace(S @ W) / input_variance
+    origin_translation = ref_mean - scale * rot_matrix @ input_mean
+
+    return (
+        [scale * (rot_matrix @ x) + origin_translation for x in input_list],
+        scale,
+        Rot3(rot_matrix),
+        origin_translation,
+    )
 
-    wTi0 = ref_list[0]
-    ref_distances = np.array([np.linalg.norm((wTi.between(wTi0)).translation()) for wTi in ref_list[1:]]) + EPSILON
 
-    # rescale poses to account for SfM scale ambiguity
-    scales = ref_distances / input_distances
-    scaling_factor = np.median(scales)
+def align_poses(input_list: List[Pose3], ref_list: List[Pose3]) -> List[Pose3]:
+    """Aligns the input list of poses to the reference list by aligning rotations followed by translations. The two
+    alignments are done independently.
+
+    Args:
+        input_list: input pose3s which need to be aligned, suppose w1Ti in world-1 frame for all frames i.
+        ref_list: reference pose3s which are target for alignment, suppose w2Ti_ in world-2 frame for all frames i.
+
+    Returns:
+         transformed pose3s which has minimum euclidean error from the reference (now living in world-2 frame).
+    """
+    # TODO: this doesn't make sense and we need to build a joint optimization.
 
-    scaled_list = [Pose3(w2Ti.rotation(), w2Ti.translation() * scaling_factor) for w2Ti in input_list]
+    # align rotations first
+    input_rotations = [x.rotation() for x in input_list]
+    ref_rotations = [x.rotation() for x in ref_list]
+    aligned_rotations = align_rotations(input_rotations, ref_rotations)
 
-    # now match origin
-    w1Ti0 = scaled_list[0]
-    i0Tw1 = w1Ti0.inverse()
-    w2Ti0_ = ref_list[0]
-    # origin transform -- map the origin of the input list to the reference list
-    w2Tw1 = w2Ti0_.compose(i0Tw1)
+    # apply the rotations to the whole pose (i.e translation too)
+    rel_rotations = [x.between(y) for x, y in zip(input_rotations, aligned_rotations)]
+    intermediate_list = [x.compose(Pose3(y, np.zeros(3))) for x, y in zip(input_list, rel_rotations)]
```

Comment:
Cant we just do 
intermediate_list = [Pose3(R, T.translation()) for R, T in zip(aligned_rotations, input_poses)]?

Reply:
I dont think I understand the purpose of intermediate_list. From what I understand, it looks like the translation in intermediate_list will be the same as in input_list. Could you please clarify the math? We can also discuss offline if you prefer.